opal/mca/threads/qthreads/threads_qthreads_module.c

```diff
@@ -34,40 +29,82 @@ struct opal_tsd_key_value {
     opal_tsd_destructor_t destructor;
 };
 
-static int opal_main_thread;
-struct opal_tsd_key_value *opal_tsd_key_values = NULL;
-static int opal_tsd_key_values_count = 0;
+/* 0: uninitialized, 1: initialized. */
+static opal_atomic_lock_t opal_thread_self_key_lock = OPAL_ATOMIC_LOCK_INIT;
+static int opal_thread_self_key_init = 0;
+static opal_tsd_key_t opal_thread_self_key;
+
+static inline void self_key_ensure_init(void)
+{
+    if (0 == opal_thread_self_key_init) {
+        /* not initialized yet. */
+        opal_atomic_lock(&opal_thread_self_key_lock);
+        /* check again. */
+        if (0 == opal_thread_self_key_init) {
+            /* This thread is responsible for initializing this key. */
+            qthread_key_create(&opal_thread_self_key, NULL);
+            opal_atomic_mb();
+            opal_thread_self_key_init = 1;
+        }
+        opal_atomic_unlock(&opal_thread_self_key_lock);
+    }
+    /* opal_thread_self_key has been already initialized. */
+}
 
 /*
  * Constructor
  */
 static void opal_thread_construct(opal_thread_t *t)
 {
+    t->t_run = 0;
+    t->t_thread_ret = 0;
 }
 
 OBJ_CLASS_INSTANCE(opal_thread_t,
                    opal_object_t,
                    opal_thread_construct, NULL);
 
+static inline aligned_t *opal_thread_get_qthreads_self(void)
+{
+    self_key_ensure_init();
+    void *ptr = qthread_getspecific(opal_thread_self_key);
+    return (aligned_t *)ptr;
+}
 
-opal_thread_t *opal_thread_get_self(void)
+static aligned_t opal_thread_qthreads_wrapper(void *arg)
 {
-    return NULL;
+    opal_thread_t *t = (opal_thread_t *)arg;
+
+    /* Register itself. */
+    self_key_ensure_init();
+    qthread_setspecific(opal_thread_self_key, t->t_thread_ret_ptr);
+
+    t->t_ret = ((void *(*)(void *))t->t_run)(t);
+    return 0;
 }
 
-bool opal_thread_self_compare(opal_thread_t *t)
+opal_thread_t *opal_thread_get_self(void)
 {
-    return OPAL_ERR_NOT_IMPLEMENTED;
+    opal_threads_ensure_init_qthreads();
```

Comment:
You are right. I removed it. This module relies on lazy initialization, so I did not need to initialize Qthreads here.

Reply:
For this line, I need `opal_threads_ensure_init_qthreads()` since a static internal function, `opal_thread_get_qthreads_self()`, calls a Qthreads function.

Reply:
What is the cost of qthread_initialize()? I wonder if it's worth calling it on each and every operation. Sooner or later qhtreads will be initialized if this module is used so why not make sure it is properly initialized from the beginning

Reply:
Reinitializing Qthreads on every operation is not cheap and it would basically result in not being able to take advantage of the benefits of using a user-level threading library rather than pthreads.

Reply:
In any case, since there's no good timing to initialize Qthreads (specifically, it is hard to forbid developers from using mutex before initializing Qthreads and after finalizing Qthreads), this check is necessary. (MPICH took another approach that is ideal for ULTs; as far as I remember, all the components in MPICH must be aware of from when to when mutex/TLS can be used, but it needs tremendous engineering efforts to maintain.)

In terms of cost, I believe `qthread_initialize()` will take this `redundant call` path, so it should not be very expensive (of course, calling an external function needs some cycles).
https://github.com/Qthreads/qthreads/blob/master/src/qthread.c#L814-L816

Maybe I miss something, however. If this function calling overhead would kill the benefits of using user-level threads, I can introduce another flag to check if Qthread has been already initialized.

Reply:
Yes, I forgot that we have that check for already being initialized.